glaze v.6.4.0 - BUG? idk

```
            /* OLD
               // booleans must be dumped using single bits
               if constexpr (has_static_size<T>) {
                  constexpr auto num_bytes = (value.size() + 7) / 8;
                  std::array<uint8_t, num_bytes> bytes{};
                  for (size_t byte_i{}, i{}; byte_i < num_bytes; ++byte_i) {
                     for (size_t bit_i = 7; bit_i < 8 && i < value.size(); --bit_i, ++i) {
                        bytes[byte_i] |= uint8_t(value[i]) << uint8_t(bit_i);
                     }
                  }
                  dump(bytes, b, ix);
               }
            */
            // FIX BEGIN
            if constexpr (has_static_size<T>) {
               constexpr auto N = std::tuple_size_v<std::decay_t<T>>;
               constexpr auto num_bytes = (N + 7) / 8;
               std::array<uint8_t, num_bytes> bytes{};
               for (size_t byte_i{}, i{}; byte_i < num_bytes; ++byte_i) {
                  for (size_t bit_i = 7; bit_i < 8 && i < N; --bit_i, ++i) {
                     bytes[byte_i] |= uint8_t(value[i]) << uint8_t(bit_i);
                  }
               }
               dump(bytes, b, ix);
            }
            // FIX END
```

So for some reason i got into weird bug with serializing ```std::array<bool, N>```
i will not pretend that i am smart enough to fix it myself and not sure if it's correct.
BUT it's compiling and working in my case.
If it is - thanks to qwen ai.

If it's not a BUG and it's my problem - explain plz whats wrong with my array.

full err log
```
[build] In file included from /workspaces/game_engine/firmware/0_external/glaze/include/glaze/glaze.hpp:36:
[build] In file included from /workspaces/game_engine/firmware/0_external/glaze/include/glaze/beve.hpp:7:
[build] In file included from /workspaces/game_engine/firmware/0_external/glaze/include/glaze/beve/ptr.hpp:7:
[build] /workspaces/game_engine/firmware/0_external/glaze/include/glaze/beve/write.hpp:554:31: error: constexpr variable 'num_bytes' must be initialized by a constant expression
[build]   554 |                constexpr auto num_bytes = (value.size() + 7) / 8;
[build]       |                               ^           ~~~~~~~~~~~~~~~~~~~~~~
[build] /workspaces/game_engine/firmware/0_external/glaze/include/glaze/beve/write.hpp:27:53: note: in instantiation of function template specialization 'glz::to<1, std::array<bool, 13>>::op<glz::opts{1, true, false, true, true, false, false, 32, 3, true, false, false, false, false, false, false, false, 0}, std::array<bool, 13> &, glz::context &, std::string &, unsigned long &>' requested here
[build]    27 |          to<BEVE, std::remove_cvref_t<T>>::template op<Opts>(std::forward<T>(value), std::forward<Ctx>(ctx),
[build]       |                                                     ^
[build] /workspaces/game_engine/firmware/0_external/glaze/include/glaze/beve/write.hpp:1111:36: note: in instantiation of function template specialization 'glz::serialize<1>::op<glz::opts{1, true, false, true, true, false, false, 32, 3, true, false, false, false, false, false, false, false, 0}, std::array<bool, 13> &, glz::context &, std::string &, unsigned long &>' requested here
[build]  1111 |                   serialize<BEVE>::op<Opts>(get_member(value, member), ctx, b, ix);
[build]       |                                    ^
[build] /workspaces/game_engine/firmware/0_external/glaze/include/glaze/beve/write.hpp:27:53: note: in instantiation of function template specialization 'glz::to<1, G::Location::Room::Facilities>::op<glz::opts{1, true, false, true, true, false, false, 32, 3, true, false, false, false, false, false, false, false, 0}, G::Location::Room::Facilities &, glz::context &, std::string &, unsigned long &>' requested here
[build]    27 |          to<BEVE, std::remove_cvref_t<T>>::template op<Opts>(std::forward<T>(value), std::forward<Ctx>(ctx),
[build]       |                                                     ^
[build] /workspaces/game_engine/firmware/0_external/glaze/include/glaze/beve/write.hpp:1111:36: note: in instantiation of function template specialization 'glz::serialize<1>::op<glz::opts{1, true, false, true, true, false, false, 32, 3, true, false, false, false, false, false, false, false, 0}, G::Location::Room::Facilities &, glz::context &, std::string &, unsigned long &>' requested here
[build]  1111 |                   serialize<BEVE>::op<Opts>(get_member(value, member), ctx, b, ix);
[build]       |                                    ^
[build] /workspaces/game_engine/firmware/0_external/glaze/include/glaze/beve/write.hpp:27:53: note: in instantiation of function template specialization 'glz::to<1, G::Location::Room::Single::Data>::op<glz::opts{1, true, false, true, true, false, false, 32, 3, true, false, false, false, false, false, false, false, 0}, G::Location::Room::Single::Data &, glz::context &, std::string &, unsigned long &>' requested here
[build]    27 |          to<BEVE, std::remove_cvref_t<T>>::template op<Opts>(std::forward<T>(value), std::forward<Ctx>(ctx),
[build]       |                                                     ^
[build] /workspaces/game_engine/firmware/0_external/glaze/include/glaze/beve/write.hpp:707:33: note: (skipping 3 contexts in backtrace; use -ftemplate-backtrace-limit=0 to see all)
[build]   707 |                serialize<BEVE>::op<Opts>(x, ctx, b, ix);
[build]       |                                 ^
[build] /workspaces/game_engine/firmware/0_external/glaze/include/glaze/beve/write.hpp:27:53: note: in instantiation of function template specialization 'glz::to<1, G::Location::Room::Manager::Data>::op<glz::opts{1, true, false, true, true, false, false, 32, 3, true, false, false, false, false, false, false, false, 0}, G::Location::Room::Manager::Data &, glz::context &, std::string &, unsigned long &>' requested here
[build]    27 |          to<BEVE, std::remove_cvref_t<T>>::template op<Opts>(std::forward<T>(value), std::forward<Ctx>(ctx),
[build]       |                                                     ^
[build] /workspaces/game_engine/firmware/0_external/glaze/include/glaze/beve/write.hpp:1111:36: note: in instantiation of function template specialization 'glz::serialize<1>::op<glz::opts{1, true, false, true, true, false, false, 32, 3, true, false, false, false, false, false, false, false, 0}, G::Location::Room::Manager::Data &, glz::context &, std::string &, unsigned long &>' requested here
[build]  1111 |                   serialize<BEVE>::op<Opts>(get_member(value, member), ctx, b, ix);
[build]       |                                    ^
[build] /workspaces/game_engine/firmware/0_external/glaze/include/glaze/core/write.hpp:25:57: note: in instantiation of function template specialization 'glz::to<1, G::DataManager::Data>::op<glz::opts{1, true, false, true, true, false, false, 32, 3, true, false, false, false, false, false, false, false, 0}, G::DataManager::Data &, glz::context &, std::string &, unsigned long &>' requested here
[build]    25 |       to<Opts.format, std::remove_cvref_t<T>>::template op<Opts>(std::forward<T>(value), ctx, buffer, ix);
[build]       |                                                         ^
[build] /workspaces/game_engine/firmware/0_external/glaze/include/glaze/core/write.hpp:70:14: note: in instantiation of function template specialization 'glz::write<glz::opts{1, true, false, true, true, false, false, 32, 3, true, false, false, false, false, false, false, false, 0}, G::DataManager::Data &, std::string, glz::context &>' requested here
[build]    70 |       return write<Opts>(std::forward<T>(value), buffer, ctx);
[build]       |              ^
[build] /workspaces/game_engine/firmware/6_g_ux_ui/2_data_manager/src/data_manager.cpp:279:59: note: in instantiation of function template specialization 'glz::write<glz::opts{1, true, false, true, true, false, false, 32, 3, true, false, false, false, false, false, false, false, 0}, G::DataManager::Data &, std::string>' requested here
[build]   279 |       case Format::GLAZE_BEVE       : glz_err_code = glz::write<glz::opts{.format = glz::BEVE, .prettify = false}>(g_buf_data, g_buf_str); break;
[build]       |                                                           ^
[build] /workspaces/game_engine/firmware/0_external/glaze/include/glaze/beve/write.hpp:554:44: note: function parameter 'value' with unknown value cannot be used in a constant expression
[build]   554 |                constexpr auto num_bytes = (value.size() + 7) / 8;
[build]       |                                            ^
[build] /workspaces/game_engine/firmware/0_external/glaze/include/glaze/beve/write.hpp:542:29: note: declared here
[build]   542 |       static void op(auto&& value, is_context auto&& ctx, auto&& b, auto&& ix)
[build]       |                             ^
[build] /workspaces/game_engine/firmware/0_external/glaze/include/glaze/beve/write.hpp:555:36: error: non-type template argument is not a constant expression
[build]   555 |                std::array<uint8_t, num_bytes> bytes{};
[build]       |                                    ^~~~~~~~~
[build] /workspaces/game_engine/firmware/0_external/glaze/include/glaze/beve/write.hpp:555:36: note: initializer of 'num_bytes' is not a constant expression
[build] /workspaces/game_engine/firmware/0_external/glaze/include/glaze/beve/write.hpp:554:31: note: declared here
[build]   554 |                constexpr auto num_bytes = (value.size() + 7) / 8;
[build]       |                               ^
[build] 2 errors generated.
```

--- a/include/glaze/beve/write.hpp
+++ b/include/glaze/beve/write.hpp
@@ -551,10 +551,11 @@ namespace glz
 
             // booleans must be dumped using single bits
             if constexpr (has_static_size<T>) {
-               constexpr auto num_bytes = (value.size() + 7) / 8;
+               constexpr auto N = std::tuple_size_v<std::decay_t<T>>;
+               constexpr auto num_bytes = (N + 7) / 8;
                std::array<uint8_t, num_bytes> bytes{};
                for (size_t byte_i{}, i{}; byte_i < num_bytes; ++byte_i) {
-                  for (size_t bit_i = 7; bit_i < 8 && i < value.size(); --bit_i, ++i) {
+                  for (size_t bit_i = 7; bit_i < 8 && i < N; --bit_i, ++i) {
                      bytes[byte_i] |= uint8_t(value[i]) << uint8_t(bit_i);
                   }
                }

--- a/tests/beve_test/beve_test.cpp
+++ b/tests/beve_test/beve_test.cpp
@@ -1473,6 +1473,64 @@ suite bitset = [] {
    };
 };
 
+suite array_bool_tests = [] {
+   "array_bool_13"_test = [] {
+      std::array<bool, 13> arr = {true, false, true, true, false, false, true, false, true, false, true, true, false};
+
+      std::string s{};
+      expect(not glz::write_beve(arr, s));
+
+      std::array<bool, 13> arr2{};
+      expect(!glz::read_beve(arr2, s));
+      expect(arr == arr2);
+   };
+
+   "array_bool_8"_test = [] {
+      std::array<bool, 8> arr = {true, false, true, false, true, false, true, false};
+
+      std::string s{};
+      expect(not glz::write_beve(arr, s));
+
+      std::array<bool, 8> arr2{};
+      expect(!glz::read_beve(arr2, s));
+      expect(arr == arr2);
+   };
+
+   "array_bool_1"_test = [] {
+      std::array<bool, 1> arr = {true};
+
+      std::string s{};
+      expect(not glz::write_beve(arr, s));
+
+      std::array<bool, 1> arr2{};
+      expect(!glz::read_beve(arr2, s));
+      expect(arr == arr2);
+   };
+};
+
+struct nested_bool_array
+{
+   int id{};
+   std::array<bool, 13> flags{};
+   std::string name{};
+
+   bool operator==(const nested_bool_array&) const = default;
+};
+
+suite nested_array_bool_tests = [] {
+   "nested_array_bool"_test = [] {
+      nested_bool_array obj{42, {true, false, true, true, false, false, true, false, true, false, true, true, false},
+                            "test"};
+
+      std::string s{};
+      expect(not glz::write_beve(obj, s));
+
+      nested_bool_array obj2{};
+      expect(!glz::read_beve(obj2, s));
+      expect(obj == obj2);
+   };
+};
+
 struct key_reflection
 {
    int i = 287;

